x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -867,6 +868,10 @@ private static Expression processBinaryOperator(BinaryOperator<?, ?, ?, ?> o) {
             return childrenChanged ? o.replaceChildren(newChildren) : o;
         }
 
+        private static boolean isSupportedCastingForBinaryComparison(DataType type) {
+            return type.isNumeric() || type == DATETIME || type == IP || type == VERSION;
```

Comment:
What's the difference between the types supported here and on the functions?
Ideally these should be in sync namely try to convert the string to all other types except numeric so to  bool, version, ip, date, geo?

Reply:
Boolean is added. 

I tried spatial types, but single-node tests failed, the error is "IllegalArgumentException: Geometry fields do not support exact searching", so I removed it from the implicit casting for binary comparison and IN list.